### ISSUE TITLE
L042 should not remove queries to CTEs when the queries reference other tables in outer query.

### ISSUE THREAD

**Issue by leo-schick** [sqlfluff/sqlfluff#3572]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

The rule L042 is a good idea I think but it does not support to config more advanced options where you - on purpose - might want to use a sub-query in a left join by using the language specific functions LATERAL (in PostgreSQL) or OUTER APPLY (in TSQL).

The Anti-pattern
``` sql
select
    a.x, a.y, b.z
from a
join (
    select x, z from b
) using(x)
```
is good, but the following pattern might be wanted, as the subquery is directly joining to the outer query:

### PostgreSQL
``` sql
select
    a.x, a.y, b.z
from a
join lateral (
    select x, z
    from b
    where b.x = a.x
    order by z
    limit 1
) on true
```

### TSQL
``` sql
select
    a.x, a.y, b.z
from a
outer apply (
    select top 1 x, y
    from b
    where b.x = a.x
    order by z
) as b
```

~I suggest to add a new config option `join-lateral` to the existing options `join` and `from`.~

Moving these queries to CTEs without refactoring the SQL will break the SQL.
I would suggest automatically excluding these subqueries from L042

### Dialect

see above

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
Why could those not be implemented as CTEs? The motivation for L042 is that CTEs are clearer.

**Comment by leo-schick** [comment]
@tunetheweb how would you define a CTE where you just want to get the first matching row?

**Comment by fmms** [comment]
@tunetheweb the LATERAL JOIN / APPLY in t-sql is used in cases where a CTE cannot be used.
Formore background have a look at https://stackoverflow.com/a/28557803 .

One is using a LATERAL join because you are having a correlated subquery. Eg. you want to just get the first matching row based on other columns in your query.

To my understanding, by definition it cannot be defined within a CTE i would say.

I was considering if it would be a good new rule to highlight when the LATERAL join is not needed.

**Comment by tunetheweb** [comment]
OK maybe your examples could be improved to actually use columns from the first queries? That's the bit I was missing.

But yes seems sensible to exclude them from L042

**Comment by fmms** [comment]
@tunetheweb give me some time, I can create a full example including some base data later today for this. Because actually in my code base I have been hit by this bug as well as it destroys code during fixing ... Will provide you something later.

**Comment by fmms** [comment]
@tunetheweb here is an example query showing the bug:

```sql
with cte_person_dates as (
    select
        name,
        cast(date as date) as date
    from (values
        (N'Person1', '20200101'),
        (N'Person1', '20150101'),
        (N'Person1', '20100101'),
        (N'Person1', '20000101')
    ) as t(name, date)
),

cte_events as (
    select
        name,
        cast(date as date) as date,
        event,
        category
    from (values
        (N'Person1', '20220710', 'job', 'new job at comp1'),
        (N'Person1', '20170101', 'job', 'new job at comp2'),
        (N'Person1', '20120101', 'job', 'new job at comp2'),
        (N'Person1', '20120101', 'education', 'finished university'),
        (N'Person1', '20060101', 'education', 'finished school')
    ) as t(name, date, category, event)
)

select
    pd.*,
    closest_event.category,
    closest_event.event
from cte_person_dates as pd
outer apply
    (select top 1
        event,
        category
     from
        cte_events as ce
     where ce.name = pd.name and ce.date <= pd.date
        order by ce.date desc) as closest_event
```

L042 now gives:
```
PS > sqlfluff fix --dialect tsql .\sqlfluff-12.sql 
==== finding fixable violations ====
== [sqlfluff-12.sql] FAIL
L:  34 | P:   5 | L042 | join_clause clauses should not contain subqueries. Use
                       | CTEs instead
```

Result is:
```sql
with cte_person_dates as (
    select
        name,
        cast(date as date) as date
    from (values
        (N'Person1', '20200101'),
        (N'Person1', '20150101'),
        (N'Person1', '20100101'),
        (N'Person1', '20000101')
    ) as t(name, date)
),

cte_events as (
    select
        name,
        cast(date as date) as date,
        event,
        category
    from (values
        (N'Person1', '20220710', 'job', 'new job at comp1'),
        (N'Person1', '20170101', 'job', 'new job at comp2'),
        (N'Person1', '20120101', 'job', 'new job at comp2'),
        (N'Person1', '20120101', 'education', 'finished university'),
        (N'Person1', '20060101', 'education', 'finished school')
    ) as t(name, date, category, event)
),

closest_event as (select top 1
        event,
        category
 from
        cte_events
 where cte_events.name = pd.name and cte_events.date <= pd.date
    order by cte_events.date desc)
select
    cte_person_dates.*,
    closest_event.category,
    closest_event.event
from cte_person_dates
outer apply
    closest_event
```

which is a corrupted query.


@leo-schick 
- is the above example showing the problem you were reporting?
- do you agree that it NEVER makes sense to move a correctly written APPLY / LATERAL operation to a CTE? In my opinion one should convert OUTER APPLY to LEFT JOIN and CROSS APPLY to INNER JOIN in case it can be correctly moved into a CTE ..

**Comment by leo-schick** [comment]
@fmms yes 🙌

i agree that we should use CET or left join, but when it is not possible, the rule should not try to force you into a CET logic (which does not work or  might result in a bad queue plan).
There might be many cases in which people should prefer a JOIN instead of APPLY/LATERAL. I suggest that anti-patterns and patterns should be drafted for this. In my case, I just APPLY/LATERAL when I do not come um with a meaningful left join (see above by using SELECT & LIMIT).

With a bad query plan I mean I could use a window function to get the number of occurrence and then use a join with filtering on „row_number“=1, but that might be more hard to read and might end up at some engines in a bad plan. (But I don’t see query plan optimizations be part of SQLfluff in this context)

**Comment by fmms** [comment]
@tunetheweb yes, I believe "L042 should not remove queries referencing other tables to CTEs" is appropriate.

**Comment by tunetheweb** [comment]
> Is there anyone planning to work on this? I don't have enough knowledge of the code base yet to tackle this one myself. @tunetheweb could you maybe tag someone who would be appropriate to take this one up/guide a PR that I'll do?

Not aware of anyone looking at this and tagging people to demand people spend their free time on something that's important to you (but possibly not them) is generally not the way open source works.

Saying that, we're usually pretty good at fixing the `fix corrupts sql` issues but think one of the reasons no progress has been made on this one, is that the issue is quite confused. The SQL in the first issue didn't clearly show the issue and the test case given in https://github.com/sqlfluff/sqlfluff/issues/3572#issuecomment-1179708478 is quite long and difficult to follow.

It also seems like there's possibly two issues in here? All in all, as someone who doesn't use this much I for one am confused.

Is the main issue that "L042 should not remove queries to CTEs when the queries reference other tables in outer query"? That makes sense to me, though think we need to update the Postgres example in [the first comment](https://github.com/sqlfluff/sqlfluff/issues/3572#issue-1298578816) to make this clear. @barrywhart any thoughts on this?

Or do we want to prevent `APPLY` and `LATERAL` queries from being pulled out (even if not referencing columns in outer query)? If so, that part might be better tackled in a separate issue to this by reopening #3688 ?

**Comment by borchero** [comment]
@tunetheweb thanks for adding your thoughts! I agree that the Postgres example could be clearer.

I'm not sure if it should be enforced to always pull out `APPLY` and `LATERAL` queries, I don't think I have sufficient experience with them.

BTW, I didn't mean to "demand" anyone to work on this, I was just eager to get this done and would appreciate some guidance in case anyone is free to do so 😄

**Comment by tunetheweb** [comment]
OK let's leave this issue as just the first one. We can open a second issue for even greater restrictions for `APPLY` and `LATERAL` in future if needs be.

I've updated the Postgres example to add a `where b.x = a.x` clause which I think makes it clearer and still keeps the intent. Let me know if that is not the case.

My experience in rule writing is limited, and having a quick look at L042 it looks quite complicated. I'd guess we either want to trim the results of `_find_nested_subqueries` to exclude subqueries with external references, or update `_calculate_fixes` to prevent those being surfaced as real lint errors. But a bit beyond me to be honest. @barrywhart may be best to advise here.

BTW, as a short term fix you can always disable L042 in your config until this is fixed.

**Comment by barrywhart** [comment]
A fix will need to determine if subqueries depend on the parents. Rules L025 and L045 do slightly similar things, so I'd begin by studying those, as well as SelectCrawler, the class used in the implementation of most rules that analyze nested queries.

**Comment by barrywhart** [comment]
@borchero: Are you volunteering to work on this? Just wanted to be sure. I may do some SQLFluff work today or this weekend. I'll choose other issues to work on if you want this one.

**Comment by borchero** [comment]
@barrywhart since I'm not at all familiar with this part of the code base yet (which seems more complicated than the dialects), I had to give this a rather low priority on my todo list, unfortunately... hence, feel free to take this one :)

**Comment by barrywhart** [comment]
I suggest using the following `ansi`-compatible query as a test case (in place of the large `tsql` queries listed earlier):
```
select
    pd.*
from cte_person_dates as pd
join
    (select *
     from cte_events as ce
     where ce.name = pd.name) as closest_event
```

We should also create a variation where the subquery uses `cte_person_dates` rather than `pd` to ensure the correlated subquery is detected and in turn prevents L042 from warning or "fixing" the query.
